tachyon/zk/plonk/vanishing/verifier_vanishing_argument.h

```diff
@@ -0,0 +1,195 @@
+// Copyright 2020-2022 The Electric Coin Company
+// Copyright 2022 The Halo2 developers
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.halo2 and the LICENCE-APACHE.halo2
+// file.
+
+#ifndef TACHYON_ZK_PLONK_VANISHING_VERIFIER_VANISHING_ARGUMENT_H_
+#define TACHYON_ZK_PLONK_VANISHING_VERIFIER_VANISHING_ARGUMENT_H_
+
+#include <utility>
+#include <vector>
+
+#include "tachyon/base/containers/container_util.h"
+#include "tachyon/crypto/transcripts/transcript.h"
+#include "tachyon/zk/base/verifier_query.h"
+#include "tachyon/zk/plonk/keys/verifying_key.h"
+
+namespace tachyon::zk {
+
+template <typename PCSTy>
+class VerifierCommitted;
+
+template <typename PCSTy>
+class VerifierConstructed;
+
+template <typename PCSTy>
+class PartiallyEvaluated;
+
+template <typename PCSTy>
+class VerifierEvaluated;
+
+template <typename PCSTy>
+class VerifierVanishingArgument {
+ public:
+  using Commitment = typename PCSTy::Commitment;
+
+  static bool ReadCommitmentsBeforeY(
+      crypto::TranscriptReader<Commitment>* transcript,
+      VerifierCommitted<PCSTy>* out) {
+    Commitment c;
+    if (!transcript->ReadPoint(&c)) return false;
+
+    *out = VerifierCommitted<PCSTy>{c};
+    return true;
+  }
+};
+
+template <typename PCSTy>
+class VerifierCommitted {
+ public:
+  using Commitment = typename PCSTy::Commitment;
+
+  constexpr VerifierCommitted() = default;
+  constexpr explicit VerifierCommitted(const Commitment& random_poly_commitment)
+      : random_poly_commitment_(random_poly_commitment) {}
+
+  constexpr Commitment& random_poly_commitment() const {
+    return random_poly_commitment_;
+  }
+
+  bool ReadCommitmentsAfterY(VerifyingKey<PCSTy> vk,
+                             crypto::TranscriptReader<Commitment>& transcript,
+                             VerifierConstructed<Commitment>* out) {
+    // Obtain a commitment to h(X) in the form of multiple pieces of degree
+    // n - 1
+    std::vector<Commitment> h_commitments;
+    h_commitments.reserve(vk.domain()->quotient_poly_degree());
+    for (Commitment& commitment : h_commitments) {
+      if (!transcript.ReadPoint(commitment)) return false;
+    }
+
+    *out = VerifierConstructed<PCSTy>{std::move(h_commitments),
+                                      random_poly_commitment_};
+    return true;
+  }
+
+ private:
+  Commitment random_poly_commitment_;
+};
+
+template <typename PCSTy>
+class VerifierConstructed {
+ public:
+  using Commitment = typename PCSTy::Commitment;
+
+  constexpr VerifierConstructed() = default;
```

Comment:
If you don't need a default constructor, please remove all these like `lookup_permuted.h`?